### DISCUSSION TITLE
Issue with creating a large number of tunnels

### DISCUSSION THREAD

**Discussion by SJin-Kim** [strongswan/strongswan#1710] [Q&A]
I am attempting to create 100,000 tunnels in strongSwan. I have noticed an issue where, despite successfully creating the first 30,000 to 40,000 tunnels, the number of half-open IKE_SAs increases, job queues become congested, and the success rate of negotiations gradually declines. Eventually, no more tunnels are created.

Additional Information:

Server Specs: 32 cores, 64GB memory
strongSwan Version: 5.8.2
strongswan.conf:
```
charon {
        threads = 32
        delete_rekeyed = yes
        install_routes = yes
        half_open_timeout = 180
        load_modular = yes
        make_before_break = yes
        reuse_ikesa = no
        ikesa_table_segments = 16
        ikesa_table_size = 32768
        plugins {
                include strongswan.d/charon/*.conf
        }
        processor {
                priority_threads{
                        low = 26
                }
        }
}
```
swanctl.conf:
```
connections {
  vpn {
  version = 2
  keyingtries = 0
  local_addrs = ****
  mobike = no
  dpd_delay = 0
  dpd_timeout = 0
  aggressive = no
  rekey_time = 28800
  over_time = 14400
  proposals = aes128-sha1-modp768
  children {
  vpn {
   esp_proposals= aes128-sha1
   start_action= none
   close_action= none
   rekey_time= 24h
   mode= tunnel
   local_ts= ****
   remote_ts= ****
   replay_window= 0
   hw_offload= yes
   inactivity= 0
  }
  }
  local {
    auth= psk
  }
  remote {
    auth= psk
  }
  }
}
secrets {
  ike {
    secret = "********"
  }
}
```
Logs: 
```25[IKE] received retransmit of request with ID 1, retransmitting response
25[CFG] no IDr configured, fall back on IP address
25[ETC]  ===> IPSEC SA CRYPTER KEY
25[KNL]    inbound : XFRM_OFFLOAD_INBOUND
25[KNL] config_hw_offload is TRUE
25[KNL]    inbound : 0
25[KNL] config_hw_offload is TRUE
13[IKE] received retransmit of request with ID 1, retransmitting response
13[CFG] no IDr configured, fall back on IP address
13[ETC]  ===> IPSEC SA CRYPTER KEY
13[KNL]    inbound : XFRM_OFFLOAD_INBOUND
13[KNL] config_hw_offload is TRUE
13[KNL]    inbound : 0
13[KNL] config_hw_offload is TRUE
27[IKE] received retransmit of request with ID 1, retransmitting response
27[IKE] received retransmit of request with ID 1, retransmitting response
27[IKE] received retransmit of request with ID 1, retransmitting response
27[CFG] no IDr configured, fall back on IP address
21[IKE] received retransmit of request with ID 1, retransmitting response
21[CFG] no IDr configured, fall back on IP address
28[IKE] received retransmit of request with ID 1, retransmitting response
28[IKE] received retransmit of request with ID 1, retransmitting response
28[CFG] no IDr configured, fall back on IP address
08[IKE] received retransmit of request with ID 1, retransmitting response
08[IKE] received retransmit of request with ID 1, retransmitting response
08[IKE] received retransmit of request with ID 1, retransmitting response
08[CFG] no IDr configured, fall back on IP address
27[ETC]  ===> IPSEC SA CRYPTER KEY
27[KNL]    inbound : XFRM_OFFLOAD_INBOUND
27[KNL] config_hw_offload is TRUE
22[IKE] received retransmit of request with ID 1, retransmitting response
22[IKE] received retransmit of request with ID 1, retransmitting response
22[CFG] no IDr configured, fall back on IP address
16[CFG] no IDr configured, fall back on IP address
21[ETC]  ===> IPSEC SA CRYPTER KEY
21[KNL]    inbound : XFRM_OFFLOAD_INBOUND
21[KNL] config_hw_offload is TRUE
13[IKE] received retransmit of request with ID 1, retransmitting response
13[IKE] received retransmit of request with ID 1, retransmitting response
13[IKE] received retransmit of request with ID 1, retransmitting response
13[IKE] received retransmit of request with ID 1, retransmitting response
13[IKE] received retransmit of request with ID 1, retransmitting response
13[IKE] received retransmit of request with ID 1, retransmitting response
13[IKE] received retransmit of request with ID 1, retransmitting response
```

**Comment by tobiasbrunner** [comment]
> I am attempting to create 100,000 tunnels in strongSwan.

In what time frame? With multiple or a single client? Are you using the `load-tester` plugin on the client?

You might want to try a newer release as there were some improvements in regards to key derivation and DoS protection.

Note that `charon.processor.priority_threads.low` doesn't have an effect (all threads basically handle low priority jobs). This feature is intended to restrict some threads to only handle higher priority jobs. See [the docs](https://docs.strongswan.org/docs/5.9/config/jobPriority.html) for details.

Did you try with more threads (some threads will always be reserved for long-running tasks, so not all will be free to handle messages)? And you should probably increase the IKE_SA table segments to roughly (has to be a power of two) match the number of threads.

The log doesn't show much. Are those retransmits received and handled by the clients? Or is the connection there already timed out? (Or is there only a single client, if so, is there perhaps a problem there?)

**Reply by SJin-Kim** [reply]
I apologize for not providing detailed information. I created multiple virtual clients using a custom load tester. These virtual clients act as initiators, and I have 100,000 of them, attempting 100 initiations per second.

I increased `priority_low_threads` because when I checked `swanctl --status`, I noticed that `half-open IKE_SAs`(more than 500) created when there were more than 30,000 tunnels. Also, there were issues with low priority jobs in the `job queues` (e.g., 0/0/0/1000). Therefore, I increased priority_low_threads to 26, the `worker threads` increased (e.g., 4/0/2/26), and the number of successful tunnels also increased.

I have also tried using more threads to 64, 128 and 512. However, with 32 CPUs, I'm not sure if having more threads would make a significant difference. Although the number of successful negotiations slightly increased, it didn't change significantly.

From the server's perspective, it seems unable to handle a large amount of authentication-related (Phase 2) information to the client. In the swanctl logs, `"successfully created shared key MAC"` is repeated multiple times in succession. After that, there is a period of approximately 10 seconds during which several phase 1 negotiations are attempted. After the negotiation attempts, the `"successfully created shared key MAC`" log appears again multiple times. This pattern repeats itself. This pattern repeated until the job queues reached 0/0/0/5000, at which point tunnel creation stopped.

**Reply by tobiasbrunner** [reply]
> and I have 100,000 of them, attempting 100 initiations per second.

Is that a typo? Did you mean "1,000 of them"? (Since you mentioned 100,000 tunnels before.)

And they do this concurrently? That is, you really attempt to create 100'000 (or even 1 million) connections within a single second?

Do you even have the hardware to handle that many crypto operations? In particular DH (even _modp768_ takes some time, it also requires entropy), and I guess the HMAC-SHA1 calculations for PSK authentication require some time too.

Also, have you tried disabling the DoS protection (`charon.dos_protection`)? (Or try a newer version, because in the version you are using there is an issue that causes the secret for COOKIES to get rotated relatively quickly, which could cause messages to get dropped.)

> Also, there were issues with low priority jobs in the `job queues` (e.g., 0/0/0/1000). Therefore, I increased priority_low_threads to 26, the `worker threads` increased (e.g., 4/0/2/26), and the number of successful tunnels also increased.

That doesn't make sense because as far as I can tell it really has no effect (the _low_ prio job queue is the last one that's checked when looking for a job, so any priority threads configured for it won't have an effect because there is no other prio/queue to be checked afterwards).



> I have also tried using more threads to 64, 128 and 512. However, with 32 CPUs, I'm not sure if having more threads would make a significant difference.

Yes, too many won't help (could even make things worse). But it probably makes sense to have at least as many free worker threads as you have cores. But it also depends on the actual tasks they perform (e.g. if threads are waiting for some I/O or lock, additional threads could possibly do some work during that time).

> In the swanctl logs, `"successfully created shared key MAC"` is repeated multiple times in succession.

What do you mean with `swanctl logs`? On the clients?  On the server? How did you configure the logging there? (In particular because that message is logged on level 2.) And if you say the message is repeated multiple times, that's for different threads I assume? No other messages logged for these threads (e.g. sent messages)?

> After that, there is a period of approximately 10 seconds during which several phase 1 negotiations are attempted.

Attempted how? You mean they fail? Up to which point do they progress?

> This pattern repeated until the job queues reached 0/0/0/5000, at which point tunnel creation stopped.

You mean nothing happens afterwards at all? Are clients still sending messages at that point?

**Reply by SJin-Kim** [reply]
> Is that a typo? Did you mean "1,000 of them"? (Since you mentioned 100,000 tunnels before.)
> And they do this concurrently? That is, you really attempt to create 100'000 (or even 1 million) connections within a single second?

I mean, I have 100,000 virtual client computers. I am attempting to create 100 connections per second(100 virtual client computers attempt to create tunnel within a single second).

> Do you even have the hardware to handle that many crypto operations? In particular DH (even _modp768_ takes some time, it also requires entropy), and I guess the HMAC-SHA1 calculations for PSK authentication require some time too.

I don't have hardware specifically designed to handle a large number of crypto operations, but currently, the server computer is not using much CPU and memory.

> Also, have you tried disabling the DoS protection (`charon.dos_protection`)? (Or try a newer version, because in the version you are using there is an issue that causes the secret for COOKIES to get rotated relatively quickly, which could cause messages to get dropped.)

That actually helped to some extent. However, after a certain number of tunnels are created, half_open_IKE is triggered, and the job queue for the 4th job increases rapidly, preventing further tunnel creation.

> That doesn't make sense because as far as I can tell it really has no effect (the _low_ prio job queue is the last one that's checked when looking for a job, so any priority threads configured for it won't have an effect because there is no other prio/queue to be checked afterwards).

I understand what you mean. I won't modify any variables related to priority anymore.

> What do you mean with `swanctl logs`? On the clients? On the server? How did you configure the logging there? (In particular because that message is logged on level 2.) And if you say the message is repeated multiple times, that's for different threads I assume? No other messages logged for these threads (e.g. sent messages)?

The logs I mentioned were from the server side, obtained by using the command `journalctl -f | grep charon`

Here is the table related to the tunnel creation attempts when 50 virtual computers try to create tunnels to the server within 1 second.

![스크린샷 2023-05-30 103748](https://github.com/strongswan/strongswan/assets/68266478/f2da120a-c32c-457c-b226-fd325a6e4cee)
![스크린샷 2023-05-30 103624](https://github.com/strongswan/strongswan/assets/68266478/6b5dc540-b9c0-45ff-881c-0d63154bdd2c)

**Reply by tobiasbrunner** [reply]
> However, after a certain number of tunnels are created, half_open_IKE is triggered

What does that mean?

> and the job queue for the 4th job increases rapidly

What does that mean? What 4th job?

Anyway, you should check what the daemon is doing exactly (log, but make sure to use sensible log levels e.g. don't just set everything to level 2).

**Reply by SJin-Kim** [reply]
In the table above, I can see that the number of tunnels successfully authenticated decreases from around 3 to 4 minutes. During this time, the accumulation of "half_open_ikesa" occurs (`IKE_SAs: 17000 total, 490 half-open`), and the job queues accumulate for AUTH-related jobs(referred to as the 4th job) with a `job queues: 0/0/0/5000`.

**Reply by tobiasbrunner** [reply]
OK, so why are they not processed? What is the daemon doing?
